projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-list-dropdown-pages-endpoint.php

```diff
@@ -0,0 +1,78 @@
+<?php // phpcs:ignore WordPress.Files.FileName.InvalidClassFileName
+
+/**
+ * List dropdown pages endpoint.
+ */
+new WPCOM_JSON_API_List_Dropdown_Pages_Endpoint(
+	array(
+
+		'description'                          => 'Get a list of pages to be displayed as options in a select-a-page-dropdown.',
+		'min_version'                          => '1.1',
+		'max_version'                          => '1.1',
+
+		'group'                                => 'posts',
+		'stat'                                 => 'posts:dropdown-pages',
+
+		'method'                               => 'GET',
+		'path'                                 => '/sites/%s/dropdown-pages/',
+		'path_labels'                          => array(
+			'$site' => '(int|string) Site ID or domain',
+		),
+
+		'allow_fallback_to_jetpack_blog_token' => true,
+
+		'example_request'                      => 'https://public-api.wordpress.com/rest/v1.1/sites/en.blog.wordpress.com/dropdown-pages/',
+	)
+);
+
+/**
+ * Endpoint class responsible for listing pages to be displayed as options in a select-a-page-dropdown.
+ *
+ * /sites/%s/dropdown-pages/ -> $blog_id
+ */
+class WPCOM_JSON_API_List_Dropdown_Pages_Endpoint extends WPCOM_JSON_API_Post_Endpoint {
+	/**
+	 * The response format.
+	 *
+	 * @var array
+	 */
+	public $response_format = array(
+		'dropdown_pages' => '(array:page) An array of page objects.',
+	);
+
+	/**
+	 * API callback.
+	 *
+	 * @param string $path - the path.
+	 * @param string $blog_id - the blog id.
+	 * @return stdClass[] $pages - array of objects containing only the ID and title fields. An empty array if no pages are found.
+	 */
+	public function callback( $path = '', $blog_id = 0 ) {
+		$blog_id = $this->api->switch_to_blog_and_validate_user( $this->api->get_blog_id( $blog_id ) );
+		if ( is_wp_error( $blog_id ) ) {
+			return $blog_id;
+		}
+
+		$pages = self::get_pages();
+		return $pages;
+	}
+
+	/**
+	 * Get all pages for the current site. The results are cached.
+	 *
+	 * @return stdClass[] $pages - array of objects containing only the ID and title fields.
+	 */
+	protected static function get_pages() {
+		global $wpdb;
+
+		$last_changed = wp_cache_get_last_changed( 'posts' );
+		$cache_key    = "get_pages:$last_changed";
+		$pages        = wp_cache_get( $cache_key, 'dropdown_pages' );
+		if ( false === $pages ) {
+			$pages = $wpdb->get_results( "SELECT {$wpdb->posts}.ID, {$wpdb->posts}.post_title as title FROM {$wpdb->posts} WHERE {$wpdb->posts}.post_type = 'page' AND {$wpdb->posts}.post_status = 'publish' ORDER BY {$wpdb->posts}.post_title ASC" );
```

Comment:
I think the `parent` is needed in order to construct the depth of the dropdown.

Hmm, is there more data that needs to sent back by this API to be able to replicate `wp_dropdown_pages()` on the client side? On that function, I see `depth` and `child_of`.

Also, it seems that there is a filter `list_pages` applied to the title of the page. Might want to filter the title through that filter. See `Walker_PageDropdown::start_el()`.

Reply:
> I think the `parent` is needed in order to construct the depth of the dropdown.

> Hmm, is there more data that needs to sent back by this API to be able to replicate `wp_dropdown_pages()` on the client side? On that function, I see `depth` and `child_of`.

I think that's a great catch! Looking at the WP Admin's Reading page and also into _Customizer → Homepage Settings_, both rely on those extra values to populate the drop-down field:

| WP Admin | Customizer |
| ------------- | ------------- |
| ![Markup on 2023-01-03 at 14:22:37](https://user-images.githubusercontent.com/25105483/210365676-fa9eaccc-0811-42b3-86b2-0df3c62fcf98.png) | ![Markup on 2023-01-03 at 14:23:05](https://user-images.githubusercontent.com/25105483/210365678-074b2585-8a2b-41a3-bdb8-6330d6be039b.png) |

---

> Also, it seems that there is a filter `list_pages` applied to the title of the page. Might want to filter the title through that filter. See `Walker_PageDropdown::start_el()`.

Looking at `Walker_PageDropdown::start_el()`, it also replaces empty page titles with page ID and `(no title)` string:

```php
		if ( '' === $title ) {
			/* translators: %d: ID of a post. */
			$title = sprintf( __( '#%d (no title)' ), $page->ID );
		}
```

The output in the drop-down field is as follows (example):
![Markup on 2023-01-03 at 14:32:02](https://user-images.githubusercontent.com/25105483/210367150-26813097-528e-4625-97f4-ac5bc196c1ba.png)

Reply:
> I think the `parent` is needed in order to construct the depth of the dropdown.

You are correct. I rushed a bit too much with the PR. Thanks for going into this for me. 🙇 

> Hmm, is there more data that needs to sent back by this API to be able to replicate `wp_dropdown_pages()` on the client side? On that function, I see `depth` and `child_of`.

I just did a more careful check. All the page properties that are required to build this type of pages dropdown tree are:
- `ID`
- `post_title`
- `post_parent`

> Also, it seems that there is a filter `list_pages` applied to the title of the page. Might want to filter the title through that filter. See `Walker_PageDropdown::start_el()`.

Uhh this is not ideal. 😞  The `list_pages` [filter](https://developer.wordpress.org/reference/hooks/list_pages/) expects the page argument to be passed along. The page is meant to be an [instance of `WP_Post`](https://developer.wordpress.org/reference/classes/wp_post/). This means that we would actually need to fetch more (probably all) post fields when making the DB query. 😕 

I found only one usage of that filter fbhepr%2Sfrnepu%3Scebwrpg%3Qjcpbz%26shyy%3Qyvfg_cntrf-og . It uses only the `post_status` property of page. We could perhaps call the filter with an incomplete `WP_Post` object instance.

What do you think @yansern ?

Reply:
> Uhh this is not ideal. 😞 The list_pages [filter](https://developer.wordpress.org/reference/hooks/list_pages/) expects the page argument to be passed along. The page is meant to be an [instance of WP_Post](https://developer.wordpress.org/reference/classes/wp_post/). This means that we would actually need to fetch more (probably all) post fields when making the DB query. 😕
>
> I found only one usage of that filter fbhepr%2Sfrnepu%3Scebwrpg%3Qjcpbz%26shyy%3Qyvfg_cntrf-og . It uses only the post_status property of page. We could perhaps call the filter with an incomplete WP_Post object instance.
>
> What do you think @yansern ?

I would wrap it in a `try...catch` and capture errors on logstash if we're going to return an incomplete `WP_Post` object.

Either that, if we use the original way of retrieving the list of pages like in the original dropdown function, although it is not highly-optimized not like the custom SQL you've created, at least it doesn't break filters that we do not potentially know where the usages are (assuming if this goes on Atomic though unlikely).

Reply:
Thanks for your suggestion @yansern . I think in that case we should probably step back and use the `get_pages` function. 

Now that I realize that pages can have children we should probably make the endpoint return a tree like data structure. Therefore I would suggest the endpoint to return the following tree-like shape of data:
```
[
  {
    id: 1,
    title: 'Hope to hope',
    children: [
      {
        id: 2,
        title: 'Always on my mind'
      }
    ]
  }
  {
    id: 3,
    title: 'Yay! Another day.'
  }
]
```

Reply:
That looks good to me. Just to clarify, the ID in the example above would be the ID of a specific post, right?

Reply:
Yes `get_pages()` function work. The API endpoint can return a tree-like data as the example you've demonstrated above. It can also:
 - return a flat array with `parent_id` & `order`, further sorting on the JS end.
 - return a flat array with `depth` (and optionally `parent_id`) and pre-sorted. one the JS end, it just needs to run through one loop to generate the dropdown.
```
[
    { id: 1, title: 'parent', depth: 0, parent_id: null }, // parent_id is not necessary but useful
    { id: 2, title: 'child', depth: 1, parent_id: 1 },
    { id: 3, title: 'grandchild', depth: 2, parent_id: 2 }
]
```

The shape of data is ultimately up to you. They all have their pros and cons.

Reply:
> Just to clarify, the ID in the example above would be the ID of a specific post, right?

Correct. 🙂